test/langtools/tools/javac/RequiredParameterFlags/ImplicitParameters.java

```diff
@@ -0,0 +1,153 @@
+/*
+ * Copyright (c) 2016, 2022, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+/*
+ * @test
+ * @bug 8292275
+ * @summary check that implicit parameter flags are available by default
+ * @library /tools/lib
+ * @modules jdk.compiler/com.sun.tools.javac.api
+ *          jdk.compiler/com.sun.tools.javac.main
+ *          jdk.compiler/com.sun.tools.javac.code
+ *          jdk.jdeps/com.sun.tools.classfile
+ * @run main ImplicitParameters
+ */
+
+import java.io.IOException;
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+import java.lang.reflect.Method;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import com.sun.tools.classfile.ClassFile;
+import com.sun.tools.classfile.ConstantPoolException;
+import com.sun.tools.classfile.MethodParameters_attribute;
+import com.sun.tools.javac.code.Flags;
+import toolbox.Assert;
+import toolbox.JavacTask;
+import toolbox.Task;
+import toolbox.TestRunner;
+import toolbox.ToolBox;
+
+public class ImplicitParameters extends TestRunner {
```

Comment:
In addition to this test, an annotation processing-based test could also be attempted here.

Reply:
Could you elaborate on that? Should I add a processor to the JavacTask in this test or write a separate one? (Maybe there is a similar test I can use as reference?)
Thank you.

Reply:
> Could you elaborate on that? Should I add a processor to the JavacTask in this test or write a separate one? (Maybe there is a similar test I can use as reference?) Thank you.

There are tests that can be used as a model around

test/langtools/tools/javac/processing/model/util/elements 

HTH

Reply:
Sorry, it looks like there is no way to access the synthetic/mandated information using the annotation processing API. Am I missing something?

Reply:
> Sorry, it looks like there is no way to access the synthetic/mandated information using the annotation processing API. Am I missing something?

For reasons of providing a minimal core language model that could be reused, less essential functionality is split in the language model API to be provided by an implementation of javax.lang.model.util.Elements.

So if you subclass JavacTestingAbstractProcessor files that test directory, you could use

eltUtils.getOrigin(elementForParameter)

to see if an element is marked as javax.lang.model.util.Enum Elements.Origin.{MANDATED, SYNTHETIC}.

Many of the annotation processing tests use a pattern of using a runtime-retention annotation to encode what the expected answer is. In this case, I'd recommend a method/constructor annotation to hold the expected origin of the parameters.

(The annotation processing API does filter out some SYNTHETIC constructs, but I don't recall if that is done at a parameter level.)

HTH

Reply:
Thanks, I started working on a test, but the results are somewhat inconsistent (no implicit parameter is present for inner class ctors, no synthetic parameters for enum ctors, parameter of valueOf is MANDATED, implicit parameters in compact record ctor are MANDATED).
I'm not sure how we want to proceed with that or if there are known issues about that, but it might be better to tackle that separately. What do you think?

Reply:
> Thanks, I started working on a test, but the results are somewhat inconsistent (no implicit parameter is present for inner class ctors, no synthetic parameters for enum ctors, parameter of valueOf is MANDATED, implicit parameters in compact record ctor are MANDATED). I'm not sure how we want to proceed with that or if there are known issues about that, but it might be better to tackle that separately. What do you think?

At least some of that behavior is expected/not surprising given the current state of javac's implementation. I linked the existing bug related bug

    JDK-8250919: Mark compiler-generated elements as mandated

to JDK-8292275 in JBS. When possible, arguably javac should present the "source view" of a construct, so it would omit, say, the synthetic parameters javac chooses to add to enum constructors.

I suggest including a test for the mandated structures that "work" and the test could be expanded as that property is tracked and used more uniformly.

Thanks and HTH